File: src/media/media.controller.ts
import {
  Controller, Post, Get, Delete, Param, Query, UseInterceptors, UploadedFile, UploadedFiles,
} from '@nestjs/common';
import { FileInterceptor, FilesInterceptor } from '@nestjs/platform-express';
import { diskStorage } from 'multer';
import { extname } from 'path';
import { v4 as uuidv4 } from 'uuid';
import { ApiTags, ApiBearerAuth, ApiConsumes, ApiBody, ApiOperation } from '@nestjs/swagger';
import { MediaService } from './media.service';
import { PaginationDto } from '../common/dto/pagination.dto';

const storage = diskStorage({
  destination: './uploads',
  filename: (_req, file, cb) => {
    const ext = extname(file.originalname);
    cb(null, `${uuidv4()}${ext}`);
  },
});

@ApiTags('Media')
@ApiBearerAuth()
@Controller('media')
export class MediaController {
  constructor(private mediaService: MediaService) {}

  @Post('upload')
  @ApiOperation({ summary: 'Upload một ảnh' })
  @ApiConsumes('multipart/form-data')
  @ApiBody({ schema: { type: 'object', properties: { file: { type: 'string', format: 'binary' } } } })
  @UseInterceptors(FileInterceptor('file', { storage }))
  upload(@UploadedFile() file: Express.Multer.File) {
    return this.mediaService.upload(file);
  }

  @Post('upload-multiple')
  @ApiOperation({ summary: 'Upload nhiều ảnh' })
  @ApiConsumes('multipart/form-data')
  @ApiBody({ schema: { type: 'object', properties: { files: { type: 'array', items: { type: 'string', format: 'binary' } } } } })
  @UseInterceptors(FilesInterceptor('files', 10, { storage }))
  uploadMultiple(@UploadedFiles() files: Express.Multer.File[]) {
    return this.mediaService.uploadMultiple(files);
  }

  @Get()
  @ApiOperation({ summary: 'Danh sách media' })
  findAll(@Query() query: PaginationDto) {
    return this.mediaService.findAll(query.page, query.limit);
  }

  @Delete(':id')
  @ApiOperation({ summary: 'Xóa media' })
  remove(@Param('id') id: string) {
    return this.mediaService.remove(+id);
  }
}

File: src/main/java/com/gese/firesecurity/media/controller/MediaController.java
package com.gese.firesecurity.media.controller;

import com.gese.firesecurity.common.dto.ApiResponse;
import com.gese.firesecurity.media.service.MediaService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

@Tag(name = "Media")
@RestController
@RequestMapping("/media")
@RequiredArgsConstructor
public class MediaController {

    private final MediaService mediaService;

    @Operation(summary = "Upload một ảnh")
    @PostMapping("/upload")
    public ResponseEntity<?> upload(@RequestParam("file") MultipartFile file) throws Exception {
        return ResponseEntity.ok(ApiResponse.ok(mediaService.upload(file)));
    }

    @Operation(summary = "Upload nhiều ảnh")
    @PostMapping("/upload-multiple")
    public ResponseEntity<?> uploadMultiple(@RequestParam("files") MultipartFile[] files) throws Exception {
        return ResponseEntity.ok(ApiResponse.ok(mediaService.uploadMultiple(files)));
    }

    @Operation(summary = "Danh sách media")
    @GetMapping
    public ResponseEntity<?> findAll(
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "20") int limit) {
        return ResponseEntity.ok(ApiResponse.ok(mediaService.findAll(page, limit)));
    }

    @Operation(summary = "Xóa media")
    @DeleteMapping("/{id}")
    public ResponseEntity<?> remove(@PathVariable Long id) {
        mediaService.remove(id);
        return ResponseEntity.ok(ApiResponse.ok("Xóa file thành công"));
    }
}
